src/Orleans/IDs/SiloAddress.cs

```diff
@@ -253,7 +253,48 @@ public bool Equals(SiloAddress other)
 
         public int CompareTo(SiloAddress other)
         {
-            return other == null ? 1 : Generation.CompareTo(other.Generation);
+            if (other == null) return 1;
+            // Compare Generation first. It gives a cheap and fast way to compare, avoiding allocations 
+            // and is also semantically meaningfull - older silos (with smaller Generation) will appear first in the comparison order.
+            // Only if Generations are the same, go on to compare Ports and IPAddress (which is more expansive to compare).
+            // Alternatively, we could compare ConsistentHashCode or UniformHashCode.
+            int comp = Generation.CompareTo(other.Generation);
+            if (comp != 0) return comp;
+
+            comp = Endpoint.Port.CompareTo(other.Endpoint.Port);
+            if (comp != 0) return comp;
+
+            return CompareIpAddresses(Endpoint.Address, other.Endpoint.Address);
+        }
+
+        // Taken from: http://www.codeproject.com/Articles/26550/Extending-the-IPAddress-object-to-allow-relative-c
+        private static int CompareIpAddresses(IPAddress one, IPAddress two)
+        {
+            int returnVal = 0;
+            if (one.AddressFamily == two.AddressFamily)
+            {
+                byte[] b1 = one.GetAddressBytes();
+                byte[] b2 = two.GetAddressBytes();
```

Comment:
ok, so do we at least want to add a code comment with @veikkoeeva's suggestion that the comparison between IPv4 and IPv6 might not be completely accurate, or alternatively just always map to IPv6 since we are not concerned about perf? I think doing either of the 2 makes sense so that it can't bite us in the future